// firebase-tools/lib/dataconnect/load.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.load = void 0;
const path = require("path");
const fileUtils = require("./fileUtils");
const types_1 = require("./types");
async function load(projectId, locationId, sourceDirectory) {
    const dataConnectYaml = await fileUtils.readDataConnectYaml(sourceDirectory);
    const serviceName = `projects/${projectId}/locations/${locationId}/services/${dataConnectYaml.serviceId}`;
    const schemaDir = path.join(sourceDirectory, dataConnectYaml.schema.source);
    const schemaGQLs = await fileUtils.readGQLFiles(schemaDir);
    const connectorInfo = await Promise.all(dataConnectYaml.connectorDirs.map(async (dir) => {
        const connectorDir = path.join(sourceDirectory, dir);
        const connectorYaml = await fileUtils.readConnectorYaml(connectorDir);
        const connectorGqls = await fileUtils.readGQLFiles(connectorDir);
        return {
            connectorYaml,
            connector: {
                name: `${serviceName}/connectors/${connectorYaml.connectorId}`,
                source: {
                    files: connectorGqls,
                },
            },
        };
    }));
    return {
        serviceName,
        sourceDirectory,
        schema: {
            name: `${serviceName}/schemas/${types_1.SCHEMA_ID}`,
            primaryDatasource: (0, types_1.toDatasource)(projectId, locationId, dataConnectYaml.schema.datasource),
            source: {
                files: schemaGQLs,
            },
        },
        dataConnectYaml,
        connectorInfo,
    };
}
exports.load = load;
