// firebase-tools/lib/dataconnect/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.toDatasource = exports.requiresVector = exports.SCHEMA_ID = void 0;
exports.SCHEMA_ID = "main";
function requiresVector(dm) {
    var _a, _b, _c, _d;
    return (_d = (_c = (_b = (_a = dm === null || dm === void 0 ? void 0 : dm.primaryDataSource) === null || _a === void 0 ? void 0 : _a.postgres) === null || _b === void 0 ? void 0 : _b.requiredExtensions) === null || _c === void 0 ? void 0 : _c.includes("vector")) !== null && _d !== void 0 ? _d : false;
}
exports.requiresVector = requiresVector;
function toDatasource(projectId, locationId, ds) {
    if (ds.postgresql) {
        return {
            postgresql: {
                database: ds.postgresql.database,
                cloudSql: {
                    instance: `projects/${projectId}/locations/${locationId}/instances/${ds.postgresql.cloudSql.instanceId}`,
                },
            },
        };
    }
    return {};
}
exports.toDatasource = toDatasource;
