// eslint-plugin-astro/lib/rules/missing-client-only-directive-value.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("../utils");
const ast_utils_1 = require("../utils/ast-utils");
const compat_1 = require("../utils/compat");
exports.default = (0, utils_1.createRule)("missing-client-only-directive-value", {
    meta: {
        docs: {
            description: "the client:only directive is missing the correct component's framework value",
            category: "Possible Errors",
            recommended: true,
        },
        schema: [],
        messages: {
            missingValue: "`client:only` directive is missing a value",
        },
        type: "problem",
    },
    create(context) {
        const sourceCode = (0, compat_1.getSourceCode)(context);
        if (!sourceCode.parserServices.isAstro) {
            return {};
        }
        function verifyDirectiveValue(attr) {
            const directiveName = (0, ast_utils_1.getAttributeName)(attr);
            if (directiveName !== "client:only")
                return;
            const directiveValue = (0, ast_utils_1.getStaticAttributeValue)(attr, context);
            if (directiveValue !== null)
                return;
            context.report({
                node: attr.name,
                messageId: "missingValue",
            });
        }
        return {
            JSXAttribute: verifyDirectiveValue,
            AstroTemplateLiteralAttribute: verifyDirectiveValue,
        };
    },
});
