// firebase-tools/lib/apphosting/config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.maybeAddSecretToYaml = exports.upsertEnv = exports.findEnv = exports.store = exports.load = exports.yamlPath = void 0;
const path_1 = require("path");
const fs_1 = require("fs");
const yaml = require("yaml");
const fs = require("../fsutils");
const prompt = require("../prompt");
const dialogs = require("./secrets/dialogs");
function yamlPath(cwd) {
    let dir = cwd;
    while (!fs.fileExistsSync((0, path_1.resolve)(dir, "apphosting.yaml"))) {
        if (fs.fileExistsSync((0, path_1.resolve)(dir, "firebase.json"))) {
            return null;
        }
        const parent = (0, path_1.dirname)(dir);
        if (parent === dir) {
            return null;
        }
        dir = parent;
    }
    return (0, path_1.resolve)(dir, "apphosting.yaml");
}
exports.yamlPath = yamlPath;
function load(yamlPath) {
    const raw = fs.readFile(yamlPath);
    return yaml.parseDocument(raw);
}
exports.load = load;
function store(yamlPath, document) {
    (0, fs_1.writeFileSync)(yamlPath, document.toString());
}
exports.store = store;
function findEnv(document, variable) {
    if (!document.has("env")) {
        return undefined;
    }
    const envs = document.get("env");
    for (const env of envs.items) {
        if (env.get("variable") === variable) {
            return env.toJSON();
        }
    }
    return undefined;
}
exports.findEnv = findEnv;
function upsertEnv(document, env) {
    if (!document.has("env")) {
        document.set("env", document.createNode([env]));
        return;
    }
    const envs = document.get("env");
    const envYaml = document.createNode(env);
    for (let i = 0; i < envs.items.length; i++) {
        if (envs.items[i].get("variable") === env.variable) {
            envs.set(i, envYaml);
            return;
        }
    }
    envs.add(envYaml);
}
exports.upsertEnv = upsertEnv;
async function maybeAddSecretToYaml(secretName) {
    const dynamicDispatch = exports;
    let path = dynamicDispatch.yamlPath(process.cwd());
    let projectYaml;
    if (path) {
        projectYaml = dynamicDispatch.load(path);
    }
    else {
        projectYaml = new yaml.Document();
    }
    if (dynamicDispatch.findEnv(projectYaml, secretName)) {
        return;
    }
    const addToYaml = await prompt.confirm({
        message: "Would you like to add this secret to apphosting.yaml?",
        default: true,
    });
    if (!addToYaml) {
        return;
    }
    if (!path) {
        path = await prompt.promptOnce({
            message: "It looks like you don't have an apphosting.yaml yet. Where would you like to store it?",
            default: process.cwd(),
        });
        path = (0, path_1.join)(path, "apphosting.yaml");
    }
    const envName = await dialogs.envVarForSecret(secretName);
    dynamicDispatch.upsertEnv(projectYaml, {
        variable: envName,
        secret: secretName,
    });
    dynamicDispatch.store(path, projectYaml);
}
exports.maybeAddSecretToYaml = maybeAddSecretToYaml;
